src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,1489 @@
+    size_t layerDeltaSize = layer->OutputSize();
+    size_t concatDeltaSize = 0;
+
+    if (childrenList[currentLayer].size() > 1)
+      layerDeltaSize *= 2;
+
+    if (parentsList[currentLayer].size() > 1)
+    {
+      concatDeltaSize = layer->InputDimensions()[0];
+      for (size_t j = 1; j < layer->InputDimensions().size(); j++)
+        concatDeltaSize *= layer->InputDimensions()[j];
+    }
+
+    // If at the last layer, we only need to add it's `concatDeltaSize`.
+    if (i == sortedNetwork.size() - 1)
+      layerDeltaSize = 0;
+
+    deltaMatrixSize += layerDeltaSize + concatDeltaSize;
+  }
+
+  if (batchSize * deltaMatrixSize == layerDeltaMatrix.n_elem)
+    return;
+
+  DeleteExtraDeltas();
+  layerDeltaMatrix = MatType(1, batchSize * deltaMatrixSize);
+  outputDeltas.clear();
+  accumulatedDeltas.clear();
+  inputDeltas.clear();
+
+  size_t offset = 0;
+  for (size_t i = 0; i < sortedNetwork.size() - 1; i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+
+    if (childrenList[currentLayer].size() > 1)
+    {
+      accumulatedDeltas.insert({ i, &layerDeltas[i] });
+      MakeAlias(*accumulatedDeltas[i], layerDeltaMatrix,
+        network[currentLayer]->OutputSize(), batchSize, offset);
+
+      outputDeltas.insert({ i, new MatType() });
+      offset += network[currentLayer]->OutputSize() * batchSize;
+    }
+    else
+      outputDeltas.insert({ i, &layerDeltas[i] });
+
+    MakeAlias(*outputDeltas[i], layerDeltaMatrix,
+       network[currentLayer]->OutputSize(), batchSize, offset);
+    offset += network[currentLayer]->OutputSize() * batchSize;
+  }
+
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    Layer<MatType>* layer = network[currentLayer];
+
+    const std::vector<size_t>& parents = parentsList[currentLayer];
+    if (parents.size() > 1)
+    {
+      inputDeltas.insert({ i, new MatType() });
+
+      size_t inputSize = layer->InputDimensions()[0];
+      for (size_t j = 1; j < layer->InputDimensions().size(); j++)
+        inputSize *= layer->InputDimensions()[j];
+
+      MakeAlias(*inputDeltas[i], layerDeltaMatrix,
+        inputSize, batchSize, offset);
+      offset += inputSize * batchSize;
+    }
+    else
+    {
+      size_t onlyParent = sortedIndices[parents.front()];
+      inputDeltas.insert({ i, outputDeltas[onlyParent]});
+    }
+  }
+  extraDeltasAllocated = true;
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::InitializeGradientPassMemory(MatType& gradient)
+{
+  size_t gradientStart = 0;
+  for (size_t i = 0; i < sortedNetwork.size(); ++i)
+  {
+    size_t currentLayer = sortedNetwork[i];
+    const size_t weightSize = network[currentLayer]->WeightSize();
+    MakeAlias(layerGradients[i], gradient, weightSize, 1, gradientStart);
+    gradientStart += weightSize;
```

Comment:
Compared to the backward pass this is a nice simple breath of fresh air. :smile: